### PR TITLE
Java Coverage with JaCoCo: part 2

### PR DESCRIPTION
**Pull request by WindLeeWT**
### Java Coverage with JaCoCo
1. Add ReportGenerator for java code coverage
2. Set BLADE_COVERAGE before running tests
3. Collecting the necessary data(source files, class files, execution data) for coverage report generation

blade_coverage_1

Report:

coverage

### CODE REVIEW

File: src/blade/configparse.py [hunk 1]
```diff
@@ -81,6 +81,7 @@ def __init__(self, current_source_dir):
             'java_test_config': {
                 'junit_libs' : [],
                 'jacoco_home' : '',
+                'coverage_report_libs' : '',
```
Comment (chen3feng) [inline src/blade/configparse.py]: 这个有main，已经不是库了，就叫coverage_reporter吧，也不是列表，应该是单数。

File: src/blade/test_runner.py [hunk 2]
```diff
@@ -233,6 +235,64 @@ def _generate_inctest_run_list(self):
...
+        Return a list of tuples(source directory, class directory, execution data)
+        for each java_test.
+            source directory: source directory of java_library target under test
+            class directory: class directory of java_library target under test
+            execution data: jacoco.exec collected by jacoco agent during testing
+        """
+        coverage_data = []
+        for key in self.tests_run_map:
+            target = self.targets[key]
+            if target.type != 'java_test':
```
Comment (chen3feng) [inline src/blade/test_runner.py]: scala_test能用jacoco吗？

Comment (chen3feng) [inline src/blade/test_runner.py]: 如果不止java能用，那么也像test一样设置个data['support_jacoco'] = True，只看这个属性，不管type，这样可扩展性好一些

File: src/blade/configparse.py (hunk 1 above)
Comment (WindLeeWT) [inline src/blade/configparse.py]: 我修改一下

File: src/blade/test_runner.py (hunk 2 above)
Comment (WindLeeWT) [inline src/blade/test_runner.py]: 目前不行，原因是：
1. 目前的scala_test通过scala启动，这里可以尝试增加javaagent选项，但是jacoco只能识别class文件，无法识别scala源文件
2. 看了一些资料，scala里面更应关注statement coverage，而不是传统的line/branch coverage，这里考虑后续使用Scoverage来做scala的覆盖率：
[One wrong option is to use tools that measure coverage looking at bytecode like JaCoCo. Even though it gives you a coverage rate number, JaCoCo knows nothing about Scala and therefore it doesn’t tell you which piece of code you forgot to cover](https://dzone.com/articles/best-code-coverage-scala)

Comment (chen3feng) [inline src/blade/test_runner.py]: ok

**Pull request merged**